File: src/main/java/kotlintraining/lesson2/optional/OptionalConversion.java
package kotlintraining.lesson2.optional;

import kotlintraining.lesson2.dto.ExtendedUserInfoDTO;
import kotlintraining.lesson2.dto.NationalIdDTO;
import kotlintraining.lesson2.dto.PrimaryUserInfoDTO;
import kotlintraining.lesson2.dto.UserInfoDTO;

import java.util.Optional;

public class OptionalConversion {

    /**
     * Return the user email if present, "none" if not
     */
    public String getUserEmail(UserInfoDTO userInfoDTO) {

        return Optional.ofNullable(userInfoDTO)
            .map(UserInfoDTO::getPrimaryUserInfo)
            .map(PrimaryUserInfoDTO::getEmail)
            .filter(email -> !email.isEmpty())
            .orElse("none");
    }

    /**
     * Return whether or not the user has an email
     */
    public boolean hasEmail(UserInfoDTO userInfoDTO) {
        // TODO: return whether or not the user has an email address
        throw new UnsupportedOperationException();
    }

    /**
     * Return an optional containing the user's bureau SSN9
     */
    public Optional<String> getBureauSsn9(UserInfoDTO userInfoDTO) {

        return Optional.ofNullable(userInfoDTO)
            .map(UserInfoDTO::getExtendedUserInfo)
            .map(ExtendedUserInfoDTO::getBureauProvidedNationalId)
            .map(NationalIdDTO::getId)
            .filter(id -> !id.isEmpty());
    }

    // Advanced

    /**
     * Return the user SSN by the following priority:
     *
     * - Bureau SSN9
     * - Bureau SSN4
     * - User SSN9
     * - User SSN4
     *
     * Any of these should be skipped if they are an null or empty string.
     */
    public String getUserSsn(UserInfoDTO userInfoDTO) {
        throw new UnsupportedOperationException();
    }
}

File: src/main/java/kotlintraining/lesson2/optional/OptionalConversion.kt
package kotlintraining.lesson2.optional

import kotlintraining.lesson2.dto.ExtendedUserInfoDTO
import kotlintraining.lesson2.dto.NationalIdDTO
import kotlintraining.lesson2.dto.PrimaryUserInfoDTO
import kotlintraining.lesson2.dto.UserInfoDTO

import java.util.Optional

class OptionalConversion {

    /**
     * Return the user email if present, "none" if not
     */
    fun getUserEmail(userInfoDTO: UserInfoDTO?): String? {

        return if( userInfoDTO?.primaryUserInfo?.email?.isEmpty() ?: true) {
            "none"
        } else {
            userInfoDTO?.primaryUserInfo?.email
        }
//        Optional.ofNullable(userInfoDTO)
//                .map<PrimaryUserInfoDTO>(Function<UserInfoDTO, PrimaryUserInfoDTO> { it.getPrimaryUserInfo() })
//                .map<String>(Function<PrimaryUserInfoDTO, String> { it.getEmail() })
//                .filter { email -> !email.isEmpty() }
//                .orElse("none")
    }

    /**
     * Return whether or not the user has an email
     */
    fun hasEmail(userInfoDTO: UserInfoDTO): Boolean {
        // TODO: return whether or not the user has an email address
        throw UnsupportedOperationException()
    }

    /**
     * Return an optional containing the user's bureau SSN9
     */
//    fun getBureauSsn9(userInfoDTO: UserInfoDTO?): String? {
//
//        //val lambda = userInfoDTO
//        return Optional.ofNullable(userInfoDTO)
//                .map<ExtendedUserInfoDTO>(Function<UserInfoDTO, ExtendedUserInfoDTO> { it.getExtendedUserInfo() })
//                .map<NationalIdDTO>(Function<ExtendedUserInfoDTO, NationalIdDTO> { it.getBureauProvidedNationalId() })
//                .map<String>(Function<NationalIdDTO, String> { it.getId() })
//                .filter { id -> !id.isEmpty() }
//    }

    // Advanced

    /**
     * Return the user SSN by the following priority:
     *
     * - Bureau SSN9
     * - Bureau SSN4
     * - User SSN9
     * - User SSN4
     *
     * Any of these should be skipped if they are an null or empty string.
     */
    fun getUserSsn(userInfoDTO: UserInfoDTO): String {
        throw UnsupportedOperationException()
    }
}
